### PR TITLE
Shift operators

### PR DESCRIPTION
**Pull request by chriseth**
Replaces https://github.com/ethereum/solidity/pull/527

### CODE REVIEW

**Comment by chriseth** [comment]
Still work in progress. Changes in Types.cpp should already be final.

**Comment by axic** [comment]
Too bad we lost history in git :(

**Comment by chriseth** [comment]
@axic there were about 20 commits covering 50 lines of code and the majority of these 50 lines need to be changed again, that's why I squashed the commits.

**Comment by pirapira** [comment]
Types.cpp looks good to me.

File: docs/types.rst [hunk 1]
```diff
@@ -136,9 +141,13 @@ Fixed-size byte arrays
 Operators:
 
 * Comparisons: ``<=``, ``<``, ``==``, ``!=``, ``>=``, ``>`` (evaluate to ``bool``)
-* Bit operators: ``&``, ``|``, ``^`` (bitwise exclusive or), ``~`` (bitwise negation)
+* Bit operators: ``&``, ``|``, ``^`` (bitwise exclusive or), ``~`` (bitwise negation), ``<<`` (left shift), ``>>`` (right shift)
 * Index access: If ``x`` is of type ``bytesI``, then ``x[k]`` for ``0 <= k < I`` returns the ``k`` th byte (read-only).
 
+The shifting operator wors with any integer type as right operand (bit will
```
Comment (federicobond) [inline docs/types.rst]: works

File: docs/types.rst [hunk 2]
```diff
@@ -51,14 +51,19 @@ Integers
 Operators:
 
 * Comparisons: ``<=``, ``<``, ``==``, ``!=``, ``>=``, ``>`` (evaluate to ``bool``)
-* Bit operators: ``&``, ``|``, ``^`` (bitwise exclusive or), ``~`` (bitwise negation)
+* Bit operators: ``&``, ``|``, ``^`` (bitwise exclusive or), ``~`` (bitwise negation), ``<<`` (left shift), ``>>`` (right shift)
```
Comment (pirapira) [inline docs/types.rst]: I think the shifts are arithmetic operators, given that they extends signs.

File: docs/types.rst [hunk 3]
```diff
@@ -136,9 +141,13 @@ Fixed-size byte arrays
...
```
Comment (pirapira) [inline docs/types.rst:144]: Here it's fine they are listed under bit operations.

File: docs/types.rst (hunk 1 above)
Comment (pirapira) [reply docs/types.rst]: also, bit -> but

File: docs/types.rst [hunk 4]
```diff
@@ -51,14 +51,19 @@ Integers
...
 
 Division always truncates (it just maps to the DIV opcode of the EVM), but it does not truncate if both
 operators are :ref:`literals<rational_literals>` (or literal expressions).
 
 Division by zero and modulus with zero throws an exception.
 
+The result of a shift operation is the type of the left operand. The
+expression ``x << y`` is equivalent to ``x * 2**y`` and ``x >> y`` is
+equivalent to ``x / 2**y``. This means that shifting negative numbers
+sign extends. Shifting by a negative amount throws an exception.
```
Comment (pirapira) [inline docs/types.rst]: Maybe `a runtime exception` as in the other paragraph.

File: docs/types.rst [hunk 5]
```diff
@@ -136,9 +141,13 @@ Fixed-size byte arrays
...
 
 * Comparisons: ``<=``, ``<``, ``==``, ``!=``, ``>=``, ``>`` (evaluate to ``bool``)
-* Bit operators: ``&``, ``|``, ``^`` (bitwise exclusive or), ``~`` (bitwise negation)
+* Bit operators: ``&``, ``|``, ``^`` (bitwise exclusive or), ``~`` (bitwise negation), ``<<`` (left shift), ``>>`` (right shift)
 * Index access: If ``x`` is of type ``bytesI``, then ``x[k]`` for ``0 <= k < I`` returns the ``k`` th byte (read-only).
 
+The shifting operator wors with any integer type as right operand (bit will
+return the type of the left operand). Shifting by a negative amount will cause
+a runtime exception.
+
```
Comment (pirapira) [inline docs/types.rst:150]: I'm wondering if shifts are counted by bytes or bits.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 6]
```diff
@@ -221,7 +237,17 @@ bool ExpressionCompiler::visit(Assignment const& _assignment)
 			// value lvalue_ref value lvalue_ref
 		}
 		m_currentLValue->retrieveValue(_assignment.location(), true);
-		appendOrdinaryBinaryOperatorCode(Token::AssignmentToBinaryOp(op), *_assignment.annotation().type);
+		utils().convertType(leftType, leftType, cleanupNeeded);
+
+		Token::Value targetOp = Token::AssignmentToBinaryOp(op);
```
Comment (pirapira) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Is this the same as `binOp`?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 7]
```diff
@@ -230,7 +256,7 @@ bool ExpressionCompiler::visit(Assignment const& _assignment)
 				m_context << swapInstruction(itemSize + lvalueSize) << Instruction::POP;
 		}
 	}
-	m_currentLValue->storeValue(*type, _assignment.location());
+	m_currentLValue->storeValue(*rightIntermediateType, _assignment.location());
```
Comment (pirapira) [inline libsolidity/codegen/ExpressionCompiler.cpp]: I think this should be the `leftType`.

Comment (pirapira) [reply libsolidity/codegen/ExpressionCompiler.cpp]: I added tests witnessing this.

**Review by pirapira — CHANGES_REQUESTED** [review]
The one about `storeValue`ed type is crucial, I think.

**Comment by chriseth** [comment]
@pirapira your tests were wrong ;-) But I think the issue was still a real issue - should be fixed now.

**Review by pirapira — APPROVED** [review]

**Comment by pirapira** [comment]
Looks good to me now.  Waiting for CIs.

**Pull request merged**